testutil/testpolylog/event.go

```diff
@@ -0,0 +1,166 @@
+// methods in a concise way.
+type EventMethodsTest struct {
+	Msg                    string
+	MsgFmt                 string
+	MsgFmtArgs             []any
+	Key                    string
+	Value                  any
+	EventMethodName        string
+	ExpectedOutputContains string
+}
+
+// RunEventMethodTests runs a set of tests for a given level.
+func RunEventMethodTests(
+	t *testing.T,
+	level polylog.Level,
+	tests []EventMethodsTest,
+	newLoggerAndOutput NewLoggerAndOutputFn,
+	newEventWithLevel NewEventWithLevelFn,
+	funcMethodEventTypeName string,
+	getExpectedLevelOutputContains func(level polylog.Level) string,
+) {
+	t.Helper()
+
+	// Title-case level string so that it can be used as the name of the
+	// method to call on the logger using reflect and for the sub-test
+	// descriptions.
+	//
+	// TODO_TECHDEBT/TODO_COMMUNITY: `strings.Title()` is deprecated. Follow
+	// migration guidance in godocs: https://pkg.go.dev/strings@go1.21.4#Title.
+	levelMethodStr := strings.Title(level.String())
+
+	for _, tt := range tests {
+		// If the test case does not specify an event method name, use the test
+		// case's key as the event method name. This is done for convenience only.
+		if tt.EventMethodName == "" {
+			tt.EventMethodName = tt.Key
+		}
+
+		var (
+			eventMethodArgs []reflect.Value
+			doneMethodName  string
+			doneMethodArgs  []reflect.Value
+		)
+
+		// Ensure that calls to #Msg(), #Msgf(), and #Send() are mutually exclusive.
+		switch {
+		case tt.Msg != "":
+			// Set up call args for polylog.Event#Msg() if tt.msg is not emtpy.
+			doneMethodName = "Msg"
+			doneMethodArgs = append(doneMethodArgs, reflect.ValueOf(tt.Msg))
+		case tt.MsgFmt != "":
+			// Set up call args for polylog.Event#Msgf() if tt.msgFmt is not emtpy.
+			doneMethodName = "Msgf"
+			doneMethodArgs = append(
+				doneMethodArgs,
+				reflect.ValueOf(tt.MsgFmt),
+				reflect.ValueOf(tt.MsgFmtArgs),
+			)
+		default:
+			// Default to calling polylog.Event#Send() if tt.msg and tt.msgFmt are
+			// both empty.
+			doneMethodName = "Send"
+		}
+
+		// If tt.EventMethodName and tt.Key are both empty, then use the done
+		// method name for the test description instead of the event method name.
+		descMethodName := tt.EventMethodName
+		if tt.EventMethodName == "" {
+			descMethodName = doneMethodName
+		}
+		testDesc := fmt.Sprintf("%s().%s()", levelMethodStr, descMethodName)
+
+		// Run this sub-test for the current level.
+		t.Run(testDesc, func(t *testing.T) {
+			logger, logOutput := newLoggerAndOutput(t, level)
+
+			// Need to use reflection in order to minimize the test code necessary
+			// to exercise all the permutations of logger level and event type methods.
+			logEvent := newEventWithLevel(t, logger, level)
+			logEventValue := reflect.ValueOf(logEvent)
+
+			// If tt.EventMethodName is not empty, build the args and call it.
+			if tt.EventMethodName != "" {
+				// Append tt.key to polylog.Event#<level>() call args.
```

Comment:
Given that `Key` and `Value` are used very similarly, why not just have an `[]Args`?

Reply:
The type of an `args` slice would have to be `[]any` which is largely antithetical to the test suite as it is exercising a bunch of type-specific methods (at each level). This would require lots of additional casting and create additional indirection.